### PR TITLE
Post build index creation

### PR DESCRIPTION
**Pull request by saurabhd336**
It may be desirable for some index types to be build post segment file has been created. This PR allows for that.
The behaviour can be achieved for a custom index being added (using index-spi) by overriding the `getIndexBuildLifecycle` method to return `BuildLifecycle.POST_SEGMENT_CREATION`

## Release notes

Adds support for building indexes post segment file creation, allowing indexes that may depend on a completed segment to be built as part of the segment creation process.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java [hunk 1]
```diff
@@ -313,6 +325,36 @@ private void handlePostCreation()
...
     }
 
+    Set<IndexType> postSegCreationIndexes = IndexService.getInstance().getAllIndexes().stream()
+        .filter(indexType -> indexType.getIndexBuildLifecycle() == IndexType.IndexBuildLifecycle.POST_SEGMENT_CREATION)
+        .collect(Collectors.toSet());
+
+    if (postSegCreationIndexes.size() > 0) {
+      // Build other indexes
+      Map<String, Object> props = new HashMap<>();
+      props.put(IndexLoadingConfig.READ_MODE_KEY, ReadMode.mmap);
```
Comment (zhtaoxiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:335]: naive question: is mmap mode alway the best choice?

**Review by zhtaoxiang — APPROVED** [review]

**Comment by zhtaoxiang** [comment]
I approved the PR by accident 😛

**Comment by Jackie-Jiang** [comment]
cc @gortiz

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java (hunk 1 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:335]: I'd say so.. This method will be called by any component when building a segment (eg by the server when commiting consuming segment, by the minions when building a segment, by spark job when building building and pushing data etc). For example the PinotRecordReader always uses mmap (used by minion jobs to read segment and build new segments out of it) https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/readers/PinotSegmentRecordReader.java#L116.

Any thoughts on this? I guess a new config can be used to specify the load type.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java [hunk 2]
```diff
@@ -38,6 +38,11 @@
  * @param <IC> the {@link IndexCreator} subclass that should be used to create indexes of this type.
  */
 public interface IndexType<C extends IndexConfig, IR extends IndexReader, IC extends IndexCreator> {
+  enum IndexBuildLifecycle {
+    DURING_SEGMENT_CREATION,
+    POST_SEGMENT_CREATION,
+    CUSTOM
```
Comment (gortiz) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java:44]: Can you add a javadoc explaining the meaning of each literal?

Comment (klsince) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java:44]: nit: just call it`BuildTime` as this enum is defined as part of IndexType

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java [hunk 3]
```diff
@@ -313,6 +325,36 @@ private void handlePostCreation()
       buildStarTreeV2IfNecessary(segmentOutputDir);
     }
 
+    Set<IndexType> postSegCreationIndexes = IndexService.getInstance().getAllIndexes().stream()
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:328]: for a bit more readability, wrap those new changes to a method like `addIndexIfNecessary`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java (hunk 1 above)
Comment (zhtaoxiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:335]: Thanks for the explanation! I think it's a good starting point if PinotRecordReader always uses mmap. We can revisit this later.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java (hunk 3 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:353]: Ack

**Review by gortiz — APPROVED** [review]

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java [hunk 4]
```diff
@@ -336,6 +349,39 @@ private void handlePostCreation()
...
+        for (IndexType indexType : postSegCreationIndexes) {
+          IndexHandler handler =
+              indexType.createIndexHandler(segmentDirectory, indexLoadingConfig.getFieldIndexConfigByColName(),
+                  _config.getSchema(), _config.getTableConfig());
+          handler.updateIndices(segmentWriter);
+        }
+      }
+    }
+  }
+
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentIndexCreationDriverImpl.java:383]: (nit) extra empty line (should probably add a check-style rule for this)

**Review by Jackie-Jiang — APPROVED** [review]
LGTM

**Pull request merged**